cluster-autoscaler/proposals/parallel_drain.md

```diff
@@ -0,0 +1,346 @@
+will be extracted from ScaleDown. Instead, NeededNodes()/UnneededNodes() methods
+will be used to fetch nodes for (un)tainting.
+
+### Algorithm State
+
+The algorithm is stateful. On each call to UpdateUnneededNodes (happening every
+CA loop iteration) the state will be updated to match the most recent cluster
+state snapshot. The most important fields to be held in algorithm state are:
+
+*   deleted\_set: set of names of nodes which are being deleted
+    (implementation-wise we will probably use NodeDeletionTracker)
+*   candidate\_set: set of names for nodes which can be deleted
+*   non\_candidate\_set: set of names of nodes for which we tried to simulate
+    drain and it failed. For each node we keep time when the simulation was
+    done.
+*   pod\_destination\_hints:
+    *   stores destination nodes computed during draining simulation. For each
+        pod from nodes in candidate\_set and deleted\_set it keeps the
+        destination node assigned during simulation.
+    *   map: source node names -> pod UID -> destination node names
+*   previous\_node\_infos
+    *   List of NodeInfo computed at the beginning of previous algorithm
+        iteration.
+*   recent\_evictions: set of pods that were recently successfully evicted from
+    a node
+
+### UpdateUnneededNodes
+
+UpdateUnneededNodes will be called from RunOnce every CA loop as it is now. On
+every call the internal state is updated to match the recent snapshot of the
+cluster state.
+
+Single loop iteration of algorithm:
+
+*   Build node\_infos list
+*   Build new algorithm state (steps below)
+
+#### State update
+
+*   Verify the pods from nodes currently being scaled down can still be
+    rescheduled
+    *   Iterate over all the nodes in deleted\_set
+    *   Validate that we can still reschedule remaining pods on other nodes in
+        the cluster
+        *   Implementation-wise this is going to reuse the logic from
+            filterOutSchedulablePodListProcessor.
+    *   Use pod\_destination\_hints and fallback to linear searching
+    *   The general algorithm here follows the same rules as "Scaledown
+        simulation loop" described in more details below.
+        *   Implementation-wise verification for nodes being scaled down and
+            looking for new candidates will be shared code.
+        *   Important differences:
+            *   We will update the
+                [pdbs\_remaining\_disruptions](#pdbs-checking) structure but
+                skip PDB checking for simulation for nodes being scaled down
+            *   We are not updating candidate\_set
+*   Verify recently evicted pods can be scheduled
+    ([details](#race-conditions-with-other-controllers))
+    *   Iterate over all the pods in recent\_evictions
+    *   If a pod was added to the list N or more seconds ago, remove it from
+        recent\_evictions
+    *   If a pod has a known owner reference, check if parent object has created
```

Comment:
We could maybe further improve this by grouping by owner ref and checking if (target_replicas - current_replicas) > recent_evictions(controller). For example if we've evicted 4 pods from a given deployment, but the deployment is only missing 1 replica we could keep just one of those on the list. 

This is just pointing out a possible future improvement. I don't think this is required from initial implementation and this comment doesn't need to be addressed in any way in this design.

Reply:
Yup, good point. Updated the doc so this idea doesn't get lost.